silkworm/sync/sentry_client.cpp

```diff
@@ -16,137 +16,300 @@
+    } catch (...) {
+        promise.set_exception(std::current_exception());
+    }
+}
+
+template <>
+awaitable<void> resolve_promise_with_awaitable_result(std::promise<void>& promise, awaitable<void> task) {
+    try {
+        co_await std::move(task);
+        promise.set_value();
+    } catch (...) {
+        promise.set_exception(std::current_exception());
+    }
+}
+
+template <typename T>
+static T sync_spawn(concurrency::TaskGroup& tasks, io_context& io_context, awaitable<T> task) {
+    std::promise<T> promise;
+    tasks.spawn(io_context, resolve_promise_with_awaitable_result(promise, std::move(task)));
+    return promise.get_future().get();
+}
+
+void SentryClient::set_status(BlockNum head_block_num, Hash head_hash, BigInt head_td, const ChainConfig& chain_config) {
+    auto fork_block_numbers = chain_config.distinct_fork_numbers();
+    auto best_block_hash = Bytes{ByteView{head_hash}};
+    auto genesis_hash = ByteView{chain_config.genesis_hash.value()};
+
+    silkworm::sentry::eth::StatusMessage status_message = {
+        0,  // the eth protocol version is replaced on the sentry end
+        chain_config.chain_id,
+        head_td,
+        best_block_hash,
+        Bytes{genesis_hash},
+        silkworm::sentry::eth::ForkId(genesis_hash, fork_block_numbers, head_block_num),
+    };
+
+    silkworm::sentry::eth::StatusData status_data = {
+        std::move(fork_block_numbers),
+        head_block_num,
+        std::move(status_message),
+    };
+
+    sync_spawn(tasks_, io_context_, sentry_client_->service()->set_status(std::move(status_data)));
+
     SILK_TRACE << "SentryClient, set_status sent";
 }
 
 void SentryClient::set_status() {
     HeaderRetrieval headers(db_access_);
-    auto [head_hash, head_td] = headers.head_hash_and_total_difficulty();
-    auto head_height = headers.head_height();
+    auto [head_height, head_hash, head_td] = headers.head_info();
     headers.close();
 
     log::Debug("Chain/db status", {"head hash", head_hash.to_hex()});
     log::Debug("Chain/db status", {"head td", intx::to_string(head_td)});
     log::Debug("Chain/db status", {"head height", std::to_string(head_height)});
 
-    set_status(head_hash, head_td, chain_config_);
+    set_status(head_height, head_hash, head_td, chain_config_);
 }
 
-void SentryClient::hand_shake() {
-    auto rpc = std::make_shared<rpc::HandShake>();
-    std::atomic_store(&handshake_, rpc);
-    exec_remotely(*handshake_);
-
-    SILK_TRACE << "SentryClient, hand_shake sent";
-    sentry::HandShakeReply reply = handshake_->reply();
+void SentryClient::handshake() {
+    auto supported_protocol = sync_spawn(tasks_, io_context_, sentry_client_->service()->handshake());
 
-    sentry::Protocol supported_protocol = reply.protocol();
-    if (supported_protocol < sentry::Protocol::ETH66) {
-        log::Critical("SentryClient") << "remote sentry do not support eth/66 protocol, stopping...";
+    if (supported_protocol < 66) {
```

Comment:
Why a nude number? I like numbers when constants do not add meaning. For example, I prefer `object.timeout(1s)` to a `object.timeout(kSomeNameTimeout)`. But here using `sentry::Protocol::ETH66` can lead our code to handle smoothly a change on the sentry::Protocol::ETH66 constant.

Reply:
Good point. I will add it to my TODO.